monix-eval/shared/src/main/scala/monix/eval/Task.scala

```diff
@@ -875,7 +875,29 @@ object Task extends TaskInstances {
     */
   def gather[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])
     (implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]): Task[M[A]] =
-    TaskGather(in)(cbf)
+    TaskGather(in.toArray, () => cbf(in))
```

Comment:
Problem with this `.toArray` is that depending on the implementation of `M[A]`:

1. this code can throw exceptions, leading to a different result than the `TaskGather` that took `in` directly (see that the original `.toArray` was inside that `try/catch`)
2. a `.toArray` on a large data-structure or on some Java Iterable that blocks threads can be an expensive operation (I know it sounds awful, but there are some truly awful Java implementations out there) and by not processing this in the context of the task, the user can get no opportunity to process it on a different thread

It's also why the second parameter is a `Function1`. My suggestion is to make `TaskGather` take a `TraversableOnce[Task[A]]` parameter and move the `.toArray` inside `TaskGather` again. Or maybe make this parameter lazy as well?